Code

tav4u opened the pull request:
As requested, incoming branch **code** is now in a pull request, **master** branch was removed and **interview** branch is now the default branch, which at this moment only contains the documentation of my proposal

kingdomik commented:
Hi @tav4u could you please add to docs example of service usage (add, list, remove, list) immediately after start up to verify it works

kingdomik commented:
why I can't add car?
```
curl -s http://localhost:8000/api/parking/slots/parked/0
[]
curl -s http://localhost:8000/api/parking/tickets/issue/0/aaa/0
{"error":"An entity does not exist"}                                                                    
```

tav4u commented:
> Hi @tav4u could you please add to docs example of service usage (add, list, remove, list) immediately after start up to verify it works

Yes, I can update the README.md with that information



> why I can't add car?
> 
> ```
> curl -s http://localhost:8000/api/parking/slots/parked/0
> []
> curl -s http://localhost:8000/api/parking/tickets/issue/0/aaa/0
> {"error":"An entity does not exist"}                                                                    
> ```

Oh, that's because the endpoint does require the site id and the fee id that was selected by the user, I still need to upload a seeder file to generate some default values on the database

kingdomik commented:
> Oh, that's because the endpoint does require the site id and the fee id that was selected by the user, I still need to upload a seeder file to generate some default values on the database
yep, without initial data I can't use it :)

tav4u commented:
> > yep, without initial data I can't use it :)
I'll provide the missing file and the instructions later today, if you have any more comments please let me know :)

tav4u commented:
To being able to add the initial set of data to the local database you need to run:

```bash
(virtual env activated) [restapi/app] $ python manage.py loaddata parking
```
To test each one of the endpoints you can use:

Add 
```bash
# first 1 correspond to site
# second 1 correspond to stand or access point
# ABCXYZ correspond to vehicle plate number
curl -s http://localhost:8000/api/parking/tickets/issue/1/ABCXYZ/1
```

Remove
```bash
# number 1 correspond to ticket 1
$ curl -s http://localhost:8000/api/parking/tickets/payment/1
```
List
```bash
# number 1 correspond to parking id 1
curl -s http://localhost:8000/api/parking/slots/parked/1
```

kingdomik commented on restapi/app/parking/views.py:
```diff
@@ -0,0 +1,101 @@
+        try:
+            ticket = Ticket.objects.get(pk=ticket_pk)
+
+            if ticket.closed:
+                serializer = TicketSerializer(ticket)
+                return Response(serializer.data)
```
Let's return error if same car removed twice and wasn't added before

tav4u replied:
Ok, now is returning a custom error if the ticket is already closed and they are trying to remove it again

kingdomik commented on restapi/app/parking/views.py:
```diff
@@ -0,0 +1,101 @@
+class ParkingAdd(APIView):
+
+    @pre_register_vehicle
+    def get(self, request, stand_pk, plate_number, stand_fee_pk, format=None):
```
Let's return error if same car parked twice and wasn't removed before

tav4u replied:
Error case was added

kingdomik commented on restapi/app/parking/ticket_algorithms.py:
```diff
@@ -0,0 +1,87 @@
+        raise NotImplementedError()
+
+
+class DefaultTicketAlgorithm(TicketAlgorithm):
```
Now we have same fee calculator for all parking stations. Let's stay it for some stations and introduce second one for another ones with same 15 free interval and hour tariff only but this time each next hour should have 10% discount up to 50%.

tav4u replied:
I missed this scenario, but, if you think this is necessary I could update my models and code to also work with this particularity

kingdomik replied:
Don't worry, you didn't missed anything :) Customer just changed requirements

tav4u replied:
Ok, changes are in progress

tav4u commented on restapi/app/parking/decorators.py:
```diff
@@ -0,0 +1,17 @@
+from .models import Vehicle
+
+
+def pre_register_vehicle(function):
+
+    @wraps(function)
+    def wrapper(request, *args, **kwargs):
+        try:
+            Vehicle.objects.get(plate_number=kwargs['plate_number'])
+        except Vehicle.DoesNotExist:
+            vehicle = Vehicle(plate_number=kwargs['plate_number'])
+            vehicle.save()
+
+        return function(request, *args, **kwargs)
+
+    return wrapper
```
Honestly I only added this because I wanted the vehicle to exist before entering this function, but, I agree with the fact that the endpoint should handle directly the resources (code was updated)

tav4u commented on restapi/app/parking/views.py:
```diff
@@ -0,0 +1,101 @@
+            return Response(serializer.data)
+
+        except ObjectDoesNotExist:
+            return Response({'error': 'An entity does not exist'}, status=409)
```
Agree, I used this exception to group [Model].DoesNotExist, so I'll change it to return the specific error messages to the API user

tav4u commented on restapi/app/parking/views.py:
```diff
@@ -0,0 +1,101 @@
+            serializer = TicketParkingSerializer(tickets, many=True)
+            return Response(serializer.data)
+        except ObjectDoesNotExist:
+            return Response({'error': 'An entity does not exist'}, status=409)
```
Yes, that's true, if for some reason a different Exception is raised, I'm not caching it, so, adding general exception case in the 3 endpoints

kingdomik replied:
we should -> we shouldn't, sorry

tav4u commented:
> a few general notes:
> 
>     * 'cycle' as part of a variable name isn't very self-descriptive. I had to look at the actual code to understand what's stored there

Would be better if I change the variable name to: `price_per_time` or `unit_price` ?

>     * the engine supports changing the fee during the parking but it has no corresponding GET request to let the user utilize it, or I missed something

That's right, at this moment I just created the 3 requested endpoints, but I could add a new one to add new fees to a ticket

>     * classess Fee, StandFee, TicketFee describe some abstractions. How these abstraction relate to each other (e.g. parent-child etc)? Why these relations are not reflected in the source code?

Fee is the base model, we cand add default Fee's to be applied to a Stand, and when the ticket is generated, I'm using TicketFee model to not lose the data, If the admin user decides to change the base Fee, that would not affect the fee applied to a ticket, so, it seems kind of transparent, but, the more tangible one is TicketFee, which is used by the algorithm calculator , also StandFee is used in the endpoints, (views.py, ticket_algorithms.py), 

> Otherwise, the code looks very good to me, especially the testing part. Thanks a lot!

Thanks!

tav4u commented:
> Could you please add logging to the service to be able debug it

Added usage of logging module, which is confured to work with the console and with a file handler, so, file debug.log will be automaticaly created

tav4u commented:
PS: I updated the README.md so the models and diagrams are now visible again

kingdomik commented:
> If the admin user decides to change the base Fee, that would not affect the fee applied to a ticket, so

Great idea!

tav4u commented:
> Now we have same fee calculator for all parking stations. Let's stay it for some stations and introduce second one for another ones with same 15 free interval and hour tariff only but this time each next hour should have 10% discount up to 50%.

Feature was introduced, test cases were added, and test data was updated to include this scenario, by default I added a promotion fee which consist as explained below:

Default fee: first 15 min are free
Initial selected fee: $ 100  / hourly fee starting with 10 % discount to 50 % with a 10% increase / hour
Car leaves after 6 hours 

 | Time | Time Range | Active Fee | Time | Discount | Total |
|------|:----------:|:----------:|:----:|:-----:|
| 0 | 1 pm - 1:15 pm | 15 min free | 15 min | 0 | $ 0 |
| 15 | 1:15 pm - 2:15 pm | Regular | 60 min |  10% |$ 90 |
| 75 | 2:15 pm - 3:15 pm | Regular | 60 min |  20%  | $ 170 |
| 135 | 3:15 pm - 4:15 pm | Regular | 60 min | 30%  |$ 240 |
| 195 | 4:15 pm - 5:15 pm | Regular | 60 min | 40%  | $ 300 |
| 255 | 5:15 pm - 6:15 pm | Regular | 60 min | 50%  | $ 400 |
| 315 | 6:15 pm - 7:15 pm | Regular | 60 min |  50% | $ 400 |

Total will be  $ 400


PS: Merge conflict is because I added a file which also lives in the main branch, I'm using this file to generate my diagrams and models, but has nothing to do with the code

dmityugov commented:
> Would be better if I change the variable name to: `price_per_time` or `unit_price` ?

yes, `price_per_time` looks a bit more self-explaining, thank you

carlosheiras17 commented:
@tav4u, first of all let me congrats you, very impressive work in such few days, the code is not difficult to read since its file structure and code design is neat, it tells you about where to put code.

just three questions, **not modifications are needed**.

- about your relational diagram: why (TicketFee - TicketFeePromotion) and (Fee - FeePromotion) looks almost repetead?  please, tell me what I'm missing.

- by current DB is possible to set different Fees for every Stand in same parking lot, is there something special/not seen by me about this abstraction?

-I'm not a design pattern expert, but since I detected some patterns like strategy (calculator injected with algorithms) and decorator here, I find very interesting to ask which design pattern fits into your ticket algorithm at the moment you apply promotions to ticket (line 59 to 65 at ticket_algorithms) from your perspective.
as soon we add more promotions, mentioned lines (if...else) will grows up with each promotion added.

kingdomik commented:
Hi @tav4u let me repeat my concern here just in case it was missed

IMPORTANT: Let's keep old calculator as is it works perfectly and can be reused later for another customer. **Please develop new one**. May be we will require bunch of new calculators soon so need flexible design to satisfy all coming customers and reduce maintenance ,may be plugins could help. Same time some of calculators could be reused or extended to don't develop from scratch for each new customer.

tav4u commented:
> @tav4u, first of all let me congrats you, very impressive work in such few days, the code is not difficult to read since its file structure and code design is neat, it tells you about where to put code.

Thanks!

> just four questions, **not modifications are needed**.
 about your relational diagram: why (TicketFee - TicketFeePromotion) and (Fee - FeePromotion) looks almost repetead?  please, tell me what I'm missing.

This approach is usually used when you are looking to preserve data regardless of modification of the original references, in my abstraction, I don't want to lose the detail of a ticket if the admin decides to change the value of the fee's or the value of the promotions
 
>     * by current DB is possible to set different Fees for every Stand in same parking lot, is there something special/not seen by me about this abstraction?


No, that fact is true, the API is allowing to handle every single stand individually even though the user final application could handle this in a different way

 
>     * I find very interesting to ask which design pattern fits into your ticket algorithm at the moment you apply promotions to ticket (line 59 to 65 at ticket_algorithms) from your perspective.
 (as soon we add more promotions, mentioned lines (if...else) will grows up with each promotion added.)

From my perspective, I already used Strategy and agree with you, I just increased the lines when added the promotion feature, I think I could use something like (Chain of responsability pattern or State pattern, which are behavioural patterns that would fit this case)
(Would you like me to do that?)
 
>     * any special reasons on why only `views` do logging?

Because I'm using MVC pattern for my rest api, I would like to mention the responsabilities of each actor to answer your question:

- The model is responsible for managing the data of the application. It receives user input from the controller.
- The view means presentation of the model in a particular 
- The controller responds to the user input and performs interactions on the data model objects. The controller receives the input, optionally validates it and then passes the input to the model.

So because the controller is the one to interact with the views and the models, is the one in charge of the logic in the application, models or view should not perform any logic related operation thus should not log information

tav4u commented:
> Hi @tav4u let me repeat my concern here just in case it was missed

> IMPORTANT: Let's keep old calculator as is it works perfectly and can be reused later for another customer. **Please develop new one**. May be we will require bunch of new calculators soon so need flexible design to satisfy all coming customers and reduce maintenance ,may be plugins could help. Same time some of calculators could be reused or extended to don't develop from scratch for each new customer.


This change is in progress, I'm rolling back the code and abstracting it to support new features without modifying the old code

tav4u commented:
> Hi @tav4u let me repeat my concern here just in case it was missed

> IMPORTANT: Let's keep old calculator as is it works perfectly and can be reused later for another customer. **Please develop new one**. May be we will require bunch of new calculators soon so need flexible design to satisfy all coming customers and reduce maintenance ,may be plugins could help. Same time some of calculators could be reused or extended to don't develop from scratch for each new customer.

The code has been updated with a new approach that will let me introduce new features without updating my main algorithm for the DefaultCalculatorAlgorithm, so, only database changes (if new fields need to be added) and class extension functionality (or plugin if I'm not wrong) are necessary, this is achieve extending the class CalculatorOperation and adding the handlers to the base, the purpose of this class is to provide single responsability function to the client class.

PS: Architecture Diagram was updated to visualize reflect this new change
